File: admin-backend/src/main/java/com/example/demo/service/UserService.java
package com.example.demo.service;

import com.example.demo.dto.*;
import com.example.demo.model.Authority;
import com.example.demo.model.User;
import com.example.demo.utils.AuthenticationProvider;

import lombok.AllArgsConstructor;

import org.springframework.core.ParameterizedTypeReference;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpMethod;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.util.List;

@Service
@AllArgsConstructor
public class UserService implements UserDetailsService {

	private static final String AUTH_API = "https://localhost:8083/auth";
	private static final String USERS_API = "https://localhost:8083/api/users";

	private final EmailService emailService;
	private final RestTemplate restTemplate;
	private final AuthenticationProvider authProvider;

	@Override
	public User loadUserByUsername(String token) {
		return this.restTemplate.exchange(
				AUTH_API, 
				HttpMethod.POST, 
				this.authProvider.getAuthEntity(new TokenDTO(token)), 
				User.class).getBody();
	}

	public AuthTokenDTO login(LoginDTO loginDTO) {
		return this.restTemplate.exchange(
				AUTH_API + "/login", 
				HttpMethod.POST, 
				this.authProvider.getAuthEntity(loginDTO), 
				AuthTokenDTO.class).getBody();
	}

	public PageDTO<UserDTO> findAll(Pageable pageable) {
		ParameterizedTypeReference<PageDTO<UserDTO>> responseType = new ParameterizedTypeReference<>() {};
		return this.restTemplate.exchange(
				String.format("%s?page=%d&size=%d", USERS_API, pageable.getPageNumber(), pageable.getPageSize()),
				HttpMethod.GET,
				this.authProvider.getAuthEntity(null),
				responseType).getBody();
	}

	public List<Authority> findAllAuthorities() {
		ParameterizedTypeReference<List<Authority>> responseType = new ParameterizedTypeReference<>() {};
		return this.restTemplate.exchange(
				USERS_API + "/authorities",
				HttpMethod.GET,
				this.authProvider.getAuthEntity(null),
				responseType).getBody();
	}

	public UserDTO create(UserDTO userDTO) {
		UserDTO user = this.restTemplate.exchange(
				USERS_API, 
				HttpMethod.POST, 
				this.authProvider.getAuthEntity(userDTO), 
				UserDTO.class).getBody();
		this.sendActivationMail(user);
		return user;
	}

	public UserDTO update(long id, UserDTO userDTO) {
		return this.restTemplate.exchange(
				USERS_API + "/" + id,
				HttpMethod.PUT,
				this.authProvider.getAuthEntity(userDTO), 
				UserDTO.class).getBody();
	}

	public void delete(long id) {
		this.restTemplate.exchange(
				USERS_API + "/" + id, 
				HttpMethod.DELETE, 
				this.authProvider.getAuthEntity(null), 
				Void.class);
	}

	public void sendActivationMail(long id) {
		UserDTO user = this.restTemplate.exchange(
				USERS_API + "/send/" + id, 
				HttpMethod.GET, 
				this.authProvider.getAuthEntity(null), 
				UserDTO.class).getBody();
		this.sendActivationMail(user);
	}

	public UserDTO getDisabled(String uuid) {
		return this.restTemplate.exchange(
				AUTH_API + "/disabled/" + uuid, 
				HttpMethod.GET, 
				this.authProvider.getAuthEntity(null), 
				UserDTO.class).getBody();
	}

	public UserDTO activate(ActivationDTO activationDTO) {
		return this.restTemplate.exchange(
				AUTH_API + "/activate", 
				HttpMethod.POST, 
				this.authProvider.getAuthEntity(activationDTO), 
				UserDTO.class).getBody();
	}

	private void sendActivationMail(UserDTO userDTO) {
		this.emailService.sendActivationLink(userDTO.getEmail(), userDTO.getFirstName(), userDTO.getActivationLink());
	}

}

File: admin-frontend/src/app/core/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { User } from '../model/user';
import { Observable, of } from 'rxjs';
import { Authority } from '../model/authority';
import { Activation } from '../model/activation';
import { Page } from '../model/page';
import { catchError, map } from 'rxjs/operators';
import { Login } from '../model/login';
import { AuthToken } from '../model/auth-token';
import { EMPTY_PAGE } from '../utils/constants';

@Injectable({
  providedIn: 'root'
})
export class UserService {

  private readonly API_PATH = 'api/users';
  private readonly AUTH_PATH = 'auth';

  constructor(private httpClient: HttpClient) { }

  login(login: Login): Observable<AuthToken> {
    return this.httpClient.post<AuthToken>(`${this.AUTH_PATH}/login`, login).pipe(
      catchError(() => of(null))
    );
  }

  findAll(page: number, size: number): Observable<Page<User>> {
    const params = new HttpParams().set('page', String(page)).set('size', String(size));
    return this.httpClient.get<Page<User>>(this.API_PATH, {params}).pipe(
      catchError(() => of(EMPTY_PAGE))
    );
  }

  findAllAuthorities(): Observable<Authority[]> {
    return this.httpClient.get<Authority[]>(`${this.API_PATH}/authorities`).pipe(
      catchError(() => of([]))
    );
  }

  save(user: User): Observable<User> {
    if (user.id){
      return this.httpClient.put<User>(`${this.API_PATH}/${user.id}`, user).pipe(
        catchError(() => of(null))
      );
    }
    return this.httpClient.post<User>(this.API_PATH, user).pipe(
      catchError(() => of(null))
    );
  }

  delete(id: number): Observable<boolean> {
    return this.httpClient.delete<null>(`${this.API_PATH}/${id}`).pipe(
      map(() => true),
      catchError(() => of(false))
    );
  }

  sendActivationMail(id: number): Observable<boolean> {
    return this.httpClient.post<null>(`${this.API_PATH}/send/${id}`, null).pipe(
      map(() => true),
      catchError(() => of(false))
    );
  }

  getDisabled(uuid: string): Observable<User> {
    return this.httpClient.get<User>(`${this.AUTH_PATH}/disabled/${uuid}`).pipe(
      catchError(() => of(null))
    );
  }

  activate(activation: Activation): Observable<User> {
    return this.httpClient.post<User>(`${this.AUTH_PATH}/activate`, activation).pipe(
      catchError(() => of(null))
    );
  }

}
